src/components/cc-logs/cc-logs.js

```diff
@@ -0,0 +1,472 @@
+    /** @type {boolean} Whether to wrap long lines */
+    this.wrapLines = false;
+
+    /** @type {Ref<HTMLElement>} */
+    this._logsRef = createRef();
+    /** @type {boolean} */
+    this._pinned = this.follow;
+    /**
+     * @type {TimestampFormatter|null} instance of the formatter to use to format timestamp.
+     * It is maintained in sync with the `timestampDisplay` property.
+     * It is `null` when the `timestampDisplay` property is `hidden`.
+     */
+    this._timestampFormatter = this._resolveTimestampFormatter();
+  }
+
+  /* region private methods */
+
+  _resolveTimestampFormatter () {
+    return this.timestampDisplay === 'hidden' ? null : new TimestampFormatter(this.timestampDisplay);
+  }
+
+  /**
+   * When logs are appended very fast, the scroll event is fired very often.
+   * We sometimes need to inhibe the onScroll callback to achieve the user action.
+   */
+  _inhibiteScrollEventTemporarily () {
+    this._scrollEventInhibited = true;
+    setTimeout(() => {
+      this._scrollEventInhibited = false;
+    }, 50);
+  }
+
+  _refreshPinnedFromScrollPosition () {
+    const element = this._logsRef.value;
+    if (element != null) {
+      this._pinned = element.scrollTop + element.offsetHeight >= element.scrollHeight - 50;
+    }
+  }
+
+  _unpin () {
+    this._pinned = false;
+    this._inhibiteScrollEventTemporarily();
+  }
+
+  _pin () {
+    this._pinned = true;
+    this._inhibiteScrollEventTemporarily();
+  }
+
+  _onScroll () {
+    if (!this.follow || this._scrollEventInhibited) {
+      return;
+    }
+    this._refreshPinnedFromScrollPosition();
+  }
+
+  /**
+   * When logs are appended very fast, the scroll event is fired very often.
+   * We need to listen to the wheel event to force unpinning.
+   */
+  _onMouseWheel (e) {
+    if (!this.follow) {
+      return;
+    }
+    if (e.wheelDeltaY > 0) {
+      this._unpin();
+    }
+  }
+
+  /* endregion */
+
+  /* region public API */
+
+  scrollToBottom () {
+    if (this.follow) {
+      this._pin();
+    }
+    else {
+      const element = this._logsRef.value;
+      if (element != null) {
+        element.scrollTop = element.scrollHeight;
+      }
+    }
+  }
+
+  /* endregion */
+
+  firstUpdated (_changedProperties) {
+    if (_changedProperties.has('follow')) {
+      this._pinned = this.follow;
+    }
+  }
+
+  willUpdate (_changedProperties) {
+    if (_changedProperties.has('timestampDisplay')) {
+      this._timestampFormatter = this._resolveTimestampFormatter();
+    }
+    if (_changedProperties.has('follow')) {
+      if (!this.follow) {
+        this._unpin();
+      }
+      else {
+        this._refreshPinnedFromScrollPosition();
+      }
+    }
+  }
+
+  render () {
+    const state = this.logs.state;
+
+    if (state === 'loading') {
+      return html`
+        <div id="loading">
+          <cc-loader></cc-loader>
+        </div>`;
+    }
+
+    if (state === 'error') {
```

Comment:
**💭 thought:** Is there a case for an error occurring after some logs have already been displayed?

Meaning, instead of hiding the logs that have been already retrieved and display a global error state, adding a special log line saying that an error occurred and that logs cannot be retrieve anymore.

Reply:
This is a very good thought. I have to think about how we could do that elegantly.

Reply:
This will be done by the `cc-logs-controller` component.